### PR TITLE
Reset media form to prevent uploading on multiple submits

### PR DESCRIPTION
**Pull request by RemcoSimonides**
Fix issue #3327

### CODE REVIEW

File: apps/catalog/catalog/src/app/dashboard/movie-tunnel/movie-tunnel.component.ts [hunk 1]
```diff
@@ -9,10 +9,8 @@ import { TunnelStep, TunnelRoot, TunnelConfirmComponent } from '@blockframes/ui/
 import { switchMap } from 'rxjs/operators';
 import { of } from 'rxjs';
 import { mergeDeep } from '@blockframes/utils/helpers';
-import { HostedMediaForm } from '@blockframes/media/form/media.form';
-import { FormGroup } from '@angular/forms';
 import { MediaService } from '@blockframes/media/+state/media.service';
-import { HostedMediaFormValue, extractMediaFromDocumentBeforeUpdate } from '@blockframes/media/+state/media.model';
+import { extractMediaFromDocumentBeforeUpdateAndUpdateMediaForms } from '@blockframes/media/+state/media.model';
```
Comment (pldespaigne) [inline apps/catalog/catalog/src/app/dashboard/movie-tunnel/movie-tunnel.component.ts:13]: don't change the function name, it's already too long ^^
and so we don't need to change it everywhere

File: apps/festival/festival/src/app/dashboard/tunnel/movie-tunnel.component.ts [hunk 2]
```diff
@@ -10,7 +10,7 @@ import { switchMap } from 'rxjs/operators';
 import { of } from 'rxjs';
 import { mergeDeep } from '@blockframes/utils/helpers';
 import { MediaService } from '@blockframes/media/+state/media.service';
-import { extractMediaFromDocumentBeforeUpdate } from '@blockframes/media/+state/media.model';
+import { extractMediaFromDocumentBeforeUpdateAndUpdateMediaForms } from '@blockframes/media/+state/media.model';
```
Comment (pldespaigne) [inline apps/festival/festival/src/app/dashboard/tunnel/movie-tunnel.component.ts:13]: same

File: libs/media/src/lib/+state/media.model.ts [hunk 3]
```diff
@@ -9,7 +9,7 @@ export * from './media.firestore';
  * and **not the front**.
  * The function also return an array of media to upload, we can then pass this array to the media service.
  */
-export function extractMediaFromDocumentBeforeUpdate(document: any) {
+export function extractMediaFromDocumentBeforeUpdateAndUpdateMediaForms(document: any, form: any) {
```
Comment (pldespaigne) [inline libs/media/src/lib/+state/media.model.ts:12]: maybe you can type `document` & `form` with something like `User | Movie | Org` & `UserForm | MovieForm | ...` ?

File: libs/media/src/lib/+state/media.model.ts [hunk 4]
```diff
@@ -46,6 +47,32 @@ function extractMediaFromDocument(document: any) {
   return medias;
 }
 
+/**
+ * Loops over form looking for mediaForms that need to be updated and then resets that form.
+ */
+function updateMediaFormInForm(form: any) {
+  if ("controls" in form) {
```
Comment (pldespaigne) [inline libs/media/src/lib/+state/media.model.ts:54]: please prefer single quotes

File: libs/organization/src/lib/organization/pages/organization/organization.component.ts [hunk 5]
```diff
@@ -4,7 +4,7 @@ import { OrganizationQuery } from '@blockframes/organization/+state/organization
 import { OrganizationService } from '@blockframes/organization/+state/organization.service';
 import { MatSnackBar } from '@angular/material/snack-bar';
 import { MediaService } from '@blockframes/media/+state/media.service';
-import { extractMediaFromDocumentBeforeUpdate } from '@blockframes/media/+state/media.model';
+import { extractMediaFromDocumentBeforeUpdateAndUpdateMediaForms } from '@blockframes/media/+state/media.model';
```
Comment (pldespaigne) [inline libs/organization/src/lib/organization/pages/organization/organization.component.ts:7]: same

File: libs/user/src/lib/auth/pages/profile-view/profile-view.component.ts [hunk 6]
```diff
@@ -11,7 +11,7 @@ import { EditPasswordForm } from '@blockframes/utils/form/controls/password.cont
 import { User } from '@blockframes/auth/+state/auth.store';
 import { DynamicTitleService } from '@blockframes/utils/dynamic-title/dynamic-title.service';
 import { MediaService } from '@blockframes/media/+state/media.service';
-import { extractMediaFromDocumentBeforeUpdate } from '@blockframes/media/+state/media.model';
+import { extractMediaFromDocumentBeforeUpdateAndUpdateMediaForms } from '@blockframes/media/+state/media.model';
```
Comment (pldespaigne) [inline libs/user/src/lib/auth/pages/profile-view/profile-view.component.ts:14]: same

File: libs/media/src/lib/+state/media.model.ts [hunk 7]
```diff
@@ -9,11 +15,13 @@ export * from './media.firestore';
  * and **not the front**.
  * The function also return an array of media to upload, we can then pass this array to the media service.
  */
-export function extractMediaFromDocumentBeforeUpdate(document: any) {
+export function extractMediaFromDocumentBeforeUpdate(document: Organization | PublicUser | Movie, form: MovieForm | OrganizationForm | ProfileForm): any {
```
Comment (pldespaigne) [inline libs/media/src/lib/+state/media.model.ts:18]: Most of the time `document` is `form.value` _(except for the movie)_
so you can simplify the function as `extractMediaFromDocumentBeforeUpdate(form: MovieForm | OrganizationForm | ProfileForm, document=form.value): any`
this way you can remove the document parameter in org & user

Comment (pldespaigne) [reply libs/media/src/lib/+state/media.model.ts:18]: also remove the return type `: any` to let typescript infer the type by itself

File: libs/media/src/lib/+state/media.model.ts [hunk 8]
```diff
@@ -1,19 +1,28 @@
...
 /**
  * This function **clean** a document from it's medias before updating it in the firestore.
  * We need to clean it because the backend functions are supposed to manage medias in the db,
  * and **not the front**.
  * The function also return an array of media to upload, we can then pass this array to the media service.
  */
-export function extractMediaFromDocumentBeforeUpdate(document: any) {
+export function extractMediaFromDocumentBeforeUpdate(
+  form: MovieForm | ProfileForm | OrganizationForm,
+  document: Organization | PublicUser | Movie = form.value): { documentToUpdate: any, mediasToUpload: HostedMediaFormValue[] } {
```
Comment (RemcoSimonides) [inline libs/media/src/lib/+state/media.model.ts:19]: I am getting an error if don't define the returned type. Do you perhaps know how we can do the trick with the <T> here? So we can define whether a Movie or Organization or PublicUser is returned?

Comment (pldespaigne) [reply libs/media/src/lib/+state/media.model.ts:19]: Good question !
I would do something like : 
```TypsScript
interface ExtarctedMediaAndDocument<T> {
  documentToUpdate: T,
  mediasToUpload: HostedFormValue[],
}

export function extractMediaFromDocumentBeforeUpdate(form: MovieForm | ProfileForm | OrganizationForm, document: Organization | PublicUser | Movie = form.value): ExtarctedMediaAndDocument<Organization | PublicUser | Movie> {
```

**Pull request merged**